### PR TITLE
Fix duplicate parents

### PR DESCRIPTION
**Pull request by fearthecowboy**
(includes regenerating tests)

### CODE REVIEW

File: modelerfour/modeler/modelerfour.ts [hunk 1]
```diff
@@ -6,6 +6,20 @@ import { Session, Channel } from '@azure-tools/autorest-extension-base';
 import { Interpretations, XMSEnum } from './interpretations';
 import { fail, minimum, pascalCase, knownMediaType, KnownMediaType } from '@azure-tools/codegen';
 
+/** adds only if the item is not in the collection already 
+ * 
+ * @note  While this isn't very efficient, it doesn't disturb the original
+ * collection, so you won't get inadvertent side effects from using Set, etc. 
+*/
+function pushDistinct<T>(targetArray: Array<T>, ...items: Array<T>): Array<T> {
```
Comment (daviwil) [inline modelerfour/modeler/modelerfour.ts:14]: I've got basically the same method later down in this file (pushIfMissing I think, I'm on my phone), is it possible to remove mine and replace usages with yours?

**Review by daviwil — APPROVED** [review]

File: modelerfour/modeler/modelerfour.ts (hunk 1 above)
Comment (fearthecowboy) [reply modelerfour/modeler/modelerfour.ts:14]: lol. Nice. 

Yeah, I can, it's ever so slightly different ( I used a rest parameter operator, but easy to make them work together)

**Review by daviwil — APPROVED** [review]

**Comment by pakrym** [comment]
I think the error might be improved a bit.

This is what I get:
```

ERROR (Modeler/DuplicateParentReference): The schema BaseClass is already referenced in an allOf by BaseClassWithDiscriminator (or one of its parents)
FATAL: Error: 2 errors occured -- cannot continue.

```

This is my swagger:
```
{
  "swagger": "2.0",
  "info": {
    "title": "Schema mapping",
    "description": "Schema mapping",
    "version": "1.0.0"
  },
  "host": "localhost:3000",
  "schemes": [
    "http"
  ],
  "produces": [
    "application/json"
  ],
  "consumes": [
    "application/json"
  ],
  "paths": {},
  "definitions": {
    "BaseClass": {
      "type": "object",
      "properties": {
        "BaseClassProperty": { "type": "string" }
      }
    },
    "BaseClassWithDiscriminator": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/BaseClass"
        }
      ],
      "properties": {
        "DiscriminatorProperty": { "type": "string" }
      },
      "discriminator": "DiscriminatorProperty",
      "required": ["DiscriminatorProperty"]
    },
    "SomeProperties": {
      "type": "object",
      "properties": {
        "SomeProperty": { "type": "string" },
        "SomeOtherProperty": { "type": "string" }
      }
    },
    "ClassThatInheritsFromBaseClassAndSomeProperties": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/BaseClass"
        },
        {
          "$ref": "#/definitions/SomeProperties"
        }
      ]
    },
    "ClassThatInheritsFromBaseClassAndSomePropertiesWithBaseClassOverride": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/BaseClass"
        },
        {
          "$ref": "#/definitions/SomeProperties"
        }
      ]
    },
    "ClassThatInheritsFromBaseClass": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/BaseClass"
        }
      ]
    },
    "ClassThatInheritsFromBaseClassWithDiscriminator": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/BaseClassWithDiscriminator"
        }
      ]
    },
    "ClassThatInheritsFromBaseClassAndBaseClassWithDiscriminator": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/BaseClass"
        },
        {
          "$ref": "#/definitions/BaseClassWithDiscriminator"
        }
      ]
    },
    "ClassThatInheritsFromBaseClassAndBaseClassWithDiscriminatorAndSomeProperties": {
      "type": "object",
      "allOf": [
        {
          "$ref": "#/definitions/BaseClass"
        },
        {
          "$ref": "#/definitions/BaseClassWithDiscriminator"
        },
        {
          "$ref": "#/definitions/SomeProperties"
        }
      ]
    }
  }
}

```

The error doesn't mention the definition that has an error neither by name nor the line number.

**Comment by pakrym** [comment]
Otherwise the fix works.

**Comment by fearthecowboy** [comment]
Lemme see if I can improve the error. 

I can't fix the lack of line numbers until we go back and put source-map tracking in.

**Comment by fearthecowboy** [comment]
@pakrym  -- I'm adding a pre-check step (ie, 'integrated linter') that runs before the modeler so we can add rules to search for bad inputs without going crazy.

I had that go thru and identify the same problem before hand, and it is able to tell you the model that actually used the `allOf` that has the upstream parent.

I'm going to close this PR and merge it with the other one and you can see if that message fits a bit better.

**Pull request closed without merging**